### PR TITLE
C api query callback

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will add a callback for a federate to respond to custom queries.

### CODE REVIEW

**Comment by phlptp** [comment]
Address #1564

File: src/helics/application_api/Federate.hpp [hunk 1]
```diff
@@ -363,6 +363,16 @@ class HELICS_CXX_EXPORT Federate {
...
     bool isQueryCompleted(query_id_t queryIndex) const;
 
+    /** supply a query callback function
+  @details the intention of the query callback is to allow federates to answer particular requests
+  through the query interface this allows other federates to make requests or queries of other
+  federates in an asynchronous fashion.
+  @param queryFunction  a function object that returns a string as a result of a query in the form
+  of const string ref. This callback will be called when a federate received a query that cannot
+  be answered that directed at a particular federate
+  */
```
Comment (kdheepak) [inline src/helics/application_api/Federate.hpp:373]: Can we add 4 spaces here before the docstring, instead of 2 spaces?

**Review by kdheepak — APPROVED** [review]

File: src/helics/shared_api_library/helicsCallbacks.h [hunk 2]
```diff
@@ -98,6 +98,44 @@ HELICS_EXPORT void helicsFilterSetCustomCallback(helics_filter filter,
...
+ * Set the data for a query callback
+ *
+ * @details There are many queries that HELICS understands directly, but it is occasionally useful to have a federate be able to respond
+ * to specific queries with answers specific to a federate
+ *
+ * @param buffer the buffer received in a helicsQueryCallback
+ * @param str pointer to the data to fill the buffer with
+ * @param strSize the size of the string
+ * @forcpponly
+ * @param[in,out] err A pointer to an error object for catching errors.
```
Comment (nightlark) [inline src/helics/shared_api_library/helicsCallbacks.h:134]: Matching other C api doc comments.

```suggestion
 * Set the data for a query callback.
 *
 * @details There are many queries that HELICS understands directly, but it is occasionally useful to have a federate be able to respond
 * to specific queries with answers specific to a federate.
 *
 * @param buffer The buffer received in a helicsQueryCallback.
 * @param str Pointer to the data to fill the buffer with.
 * @param strSize The size of the string.
 * @forcpponly
 * @param[in,out] err A pointer to an error object for catching errors.
```

File: src/helics/shared_api_library/helicsCallbacks.h [hunk 3]
```diff
@@ -98,6 +98,44 @@ HELICS_EXPORT void helicsFilterSetCustomCallback(helics_filter filter,
...
+ * to specific queries with answers specific to a federate
+ *
+ * @param fed The federate to set the callback for
+ * @param queryAnswer A callback with signature const char *(const char *query, int querySize,int *answerSize, void *userdata);
+ *                 The function arguments are the query string requesting an answer along with its size, the string is not guaranteed to be
+ * null terminated answerSize is an outputParameter intended to filled out by the userCallback and should contain the length of the return
+ * string. the return pointer can be NULL if no answer is given and HELICS will generate the appropriate response.
+ * @param userdata A pointer to user data that is passed to the function when executing.
+ * @forcpponly
+ * @param[in,out] err A pointer to an error object for catching errors.
```
Comment (nightlark) [inline src/helics/shared_api_library/helicsCallbacks.h:114]: Matching other C api doc comments.

```suggestion
 * Set callback for queries executed against a federate.
 *
 * @details There are many queries that HELICS understands directly, but it is occasionally useful to have a federate be able to respond
 * to specific queries with answers specific to a federate.
 *
 * @param fed The federate to set the callback for.
 * @param queryAnswer A callback with signature const char *(const char *query, int querySize,int *answerSize, void *userdata);
 *                 The function arguments are the query string requesting an answer along with its size, the string is not guaranteed to be
 * null terminated answerSize is an outputParameter intended to filled out by the userCallback and should contain the length of the return
 * string. The return pointer can be NULL if no answer is given and HELICS will generate the appropriate response.
 * @param userdata A pointer to user data that is passed to the function when executing.
 * @forcpponly
 * @param[in,out] err A pointer to an error object for catching errors.
```

File: src/helics/core/CommonCore.cpp [hunk 4]
```diff
@@ -2064,8 +2064,7 @@ std::string CommonCore::federateQuery(const FederateState* fed, const std::strin
         return filteredEndpointQuery(fed);
     }
     if ((queryStr == "queries") || (queryStr == "available_queries")) {
-        return std::string(
-                   "[exists;isinit;state;global_state;version;queries;filtered_endpoints;current_time;") +
+        return std::string("[exists;isinit;state;version;queries;filtered_endpoints;") +
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/core/CommonCore.cpp:2067]: This list is queries about a federate that are answered by a core, so current time in this list would be duplicative as it is also in the federate list

**Review by nightlark — APPROVED** [review]

**Pull request merged**